### ISSUE TITLE
TDN training does not behave as expected

### ISSUE THREAD

**Issue by gabrielecaddeo** [facebookresearch/MidasTouch#4]
Hi,
Thanks again for sharing with us your amazing work.

Following the discussion of the issue #1, I managed to retrain the TDN but the results are not as expected. Indeed, the codebook is quite different, as it can be seen in this table:

| Input | Provided weights | Retrain weights | 
:-------------------------:|:-------------------------:|:-------------------------:
004_sugar_box |![](https://user-images.githubusercontent.com/49904924/210998995-67f77ab9-e4b2-432d-b2ee-fb8676e04c83.png) |![](https://user-images.githubusercontent.com/49904924/210998933-046dbd93-5e9d-4b90-a672-a9ee6f2d291b.png) 
005_tomato_soup_can |![](https://user-images.githubusercontent.com/49904924/210999223-d4275d5b-c2ad-4e2b-b37c-e934e7cf1e11.png) |![](https://user-images.githubusercontent.com/49904924/210999157-c5c9867c-e8f5-4687-99dc-3479aec58b65.png)

I checked also the output mask to investigate the problem by using the same input images for the 2 cases:

 | Input | Provided weights | Retrain weights | 
:-------------------------:|:-------------------------:|:-------------------------:
![](https://user-images.githubusercontent.com/49904924/210603858-51dccef3-12b4-470a-b8f6-28b7bca2cb71.png) |![](https://user-images.githubusercontent.com/49904924/210609770-24cbeff5-67a5-48fb-982b-54f32b4039e9.png) |![](https://user-images.githubusercontent.com/49904924/210609841-7b046a47-4cdd-4383-b486-266f6ab2a863.png) 
![](https://user-images.githubusercontent.com/49904924/210603738-c9b00a7b-f6a5-423a-872e-e153211c4601.png) |![](https://user-images.githubusercontent.com/49904924/210609774-12d98e1b-57e2-49ed-9acc-9f4f3821186f.png) | ![](https://user-images.githubusercontent.com/49904924/210609844-743cdba8-fb5f-467b-b098-8156e730baa8.png) 
![](https://user-images.githubusercontent.com/49904924/210603742-d2d78ac4-2916-47f0-9ae5-b4f4a91d1cfe.png) |![](https://user-images.githubusercontent.com/49904924/210609778-68b8406a-c124-48b8-88a5-71e823a3b796.png) | ![](https://user-images.githubusercontent.com/49904924/210609846-fa7aa58b-cc9a-4a25-9f5c-86402bed0cf8.png) 
![](https://user-images.githubusercontent.com/49904924/210603744-45be1d2e-8c38-4171-8f1a-91b8dafaae80.png) |![](https://user-images.githubusercontent.com/49904924/210609782-ed8708f5-446e-4389-92c9-2e813fe26c9e.png) | ![](https://user-images.githubusercontent.com/49904924/210609847-14287cf1-ab1c-4d1b-827a-9ea9200c8950.png) 
![](https://user-images.githubusercontent.com/49904924/210603746-2a9f50fd-94e2-42d6-9781-752118fb079f.png) |![](https://user-images.githubusercontent.com/49904924/210609784-94a62cb5-316e-4bd8-a6e8-a068a51868fc.png) | ![](https://user-images.githubusercontent.com/49904924/210609848-6e9b37ec-8c45-464b-940f-2d859c4eed03.png) 
![](https://user-images.githubusercontent.com/49904924/210552591-551c5036-42b9-4f37-9b4d-6a8e9b139206.png) |![](https://user-images.githubusercontent.com/49904924/210609789-9cec8e47-810c-4f71-812c-2a30fd8b01ee.png) | ![](https://user-images.githubusercontent.com/49904924/210609850-dc0caad0-5cec-4e48-bcd1-c8c7917f1493.png) 
![](https://user-images.githubusercontent.com/49904924/210603753-19a022fa-c38f-4bed-a168-69ea8b4eee34.png) |![](https://user-images.githubusercontent.com/49904924/210609790-f57a4bec-7776-4e3a-8edd-b603deb54906.png) | ![](https://user-images.githubusercontent.com/49904924/210609854-f6d1303c-8b2d-4555-8115-217b0dd7b425.png) 
![](https://user-images.githubusercontent.com/49904924/210603756-802be5e6-6d57-4e2e-87c0-738e5387f18e.png) |![](https://user-images.githubusercontent.com/49904924/210609794-59147cc9-2432-4343-8ebb-608c93091f6f.png) | ![](https://user-images.githubusercontent.com/49904924/210609857-61fb9bb4-c62d-4000-8b05-8d25a790a32f.png) 

I used the default configuration values, 10k images for each of the 40 training objects and the NYU weights as pretrain. 

I don't know if I am missing something, do you have any idea why I get this behavior? Thanks!

**Comment by suddhu** [comment]
Hi @gabrielecaddeo!

Thank you for the detailed description. First, it does appear as though the re-trained TDN is not performing as well. This can lead to incorrect pointclouds, and therefore inaccurate embeddings. 

- Are the images you show below from TACTO or the real sensor? 
- Did the training procedure ([train.py](https://github.com/facebookresearch/MidasTouch/blob/main/midastouch/contrib/tdn_fcrn/train.py)) converge to a low depth RMSE? 
- If there is a mismatch between the background used in training, and that of testing, it can also lead to errors. 

I think once we resolve your bug in the TDN, we can have a look at the embedding part. Let me know if you have any further information to report too, thanks for all your interest!

**Comment by gabrielecaddeo** [comment]
Hi @suddhu ! Yes, the problem seems in the TDN training.

>     * Are the images you show below from TACTO or the real sensor?
The images I showed are an example of images generated from TACTO by touching the 004_sugar_box with the backgrounds you provided in the repo. The input images I loaded in the previous comment had the channel inverted because I saved them after [this](https://github.com/facebookresearch/MidasTouch/blob/13f0887f1676b6a1d03e7e7bbdee8c36c11bbe9e/midastouch/tactile_tree/build_codebook.py#L85) line of code. These are the real input, I think now you can recognize your background.
 | Input | 
:-------------------------:
![](https://user-images.githubusercontent.com/49904924/211785804-6053affb-7c4b-44b8-b586-9692b66bcdd9.png) 
![](https://user-images.githubusercontent.com/49904924/211785810-af59b6d6-346c-4b12-8ce0-bda711039f3b.png) 
![](https://user-images.githubusercontent.com/49904924/211785811-f5de5391-ddc0-4ffb-8e10-691ea9ec2a04.png) 




>     * Did the training procedure ([train.py](https://github.com/facebookresearch/MidasTouch/blob/main/midastouch/contrib/tdn_fcrn/train.py)) converge to a low depth RMSE?

I do not have the plot of the loss for the entire training, but I have the numbers of the last checkpoint saved.
The last checkpoint saved was on epoch 93/100. This means the next epochs had a higher error.
These are the details of the epoch 93:
- Epoch error of 23.428 pixel RMSE.
- Validation error of 20.007 pixel RMSE, with a previous validation error of 20.378 pixel RMSE.
- The test error was 0.599 mm RMSE

>     * If there is a mismatch between the background used in training, and that of testing, it can also lead to errors.

There was no mismatch, as you can see in the first part of this comment that I used your background for the training and the testing.

In order to be 100% clear, I generated the training data by running the [generate_data.py](https://github.com/facebookresearch/MidasTouch/blob/main/midastouch/data_gen/generate_data.py) script.
Then I run the [data_to_txt.py](https://github.com/facebookresearch/MidasTouch/blob/main/midastouch/contrib/tdn_fcrn/data/data_to_txt.py) script to create the labels. Then I run the training with the [train.py](https://github.com/facebookresearch/MidasTouch/blob/main/midastouch/contrib/tdn_fcrn/train.py) script.
I always used the default configuration files. The only thing I changed was to set [this](https://github.com/facebookresearch/MidasTouch/blob/13f0887f1676b6a1d03e7e7bbdee8c36c11bbe9e/midastouch/contrib/tdn_fcrn/config/train.yaml#L11) flag to False to load the NYU weights, since I was running a new training.

**Comment by suddhu** [comment]
Interesting, this could probably be an issue with training convergence -  for our TDN we report 0.135 mm test error, and the pixel RMSE is lower than 20px too. It's probably that the default learning rate is too small. It could also be worth overfitting to a smaller tactile dataset, and checking if you get both qualitative (visualize depth and contact masks) and quantitative (RMSE) convergence. 

Apologies for the late response, if you email me we could set up a meeting over the next week so we can iterate on this faster!

**Comment by gabrielecaddeo** [comment]
Hi @suddhu , I wrote you an email to your institutional address, suddhu@cmu.edu. Thank you

**Comment by suddhu** [comment]
Hi @gabrielecaddeo! 

As we discussed in the meeting I ran the TDN training procedure from scratch (for a few epochs). I will share my summary and data below, in case it may benefit others too: 

1. `python midastouch/data_gen/generate_data.py method=train_data` with `train_data.yaml` set to: 
```yaml
obj_class: ycb_train
obj_model: 035_power_drill
sampling : random # random, random+edges, traj, manual
save_path : ./train_data
noise:
 sig_r: 1  # degrees
 sig_t: 5e-4  # m

num_samples : 10000
total_length : 0.5

render:
 pixmm : 0.03
 width : 240
 height : 320
 cam_dist : 0.022
 shear_mag : 5.0
 pen :
   min : 0.0005
   max : 0.001
 randomize : True
```
I've uploaded the entire training dataset here ([Jan31_train_data.zip](https://drive.google.com/file/d/1ZSqL0TbmWl7wzXlnGjRbowf3GZXDfrkw/view?usp=sharing)). 

---
2. `python midastouch/contrib/tdn_fcrn/data/data_to_txt.py` after setting the appropriate folder path 
---
3. `python midastouch/contrib/tdn_fcrn/train.py` starting from the pre-trained NYU weights. `train.yaml` was set to: 
```yaml
batch_size : 50
lr : 1.0e-4
max_epochs : 100
resume_from_file : False
data_file_path : data
init_weights : NYU_ResNet-UpProj.npy
checkpoint_weights : tdn_weights.pth.tar
```
I trained for 2 epochs with `lr: 1e-4` and 3rd epoch with `lr: 1e-5`.  Here are some excerpts from the command line: 
```bash
Before train error: 64.064 pixel RMSE
Starting train epoch 1 / 100
...
Starting train epoch 3 / 100
RMSE pixel loss: 10.79
...
Epoch error: 10.791 pixel RMSE
Validation error: 7.275 pixel RMSE, Best validation error: 9.704 pixel RMSE
Test error: 0.214 mm RMSE
```

You can see sample image/heightmap/gt from the [training](https://drive.google.com/file/d/1L91eWVJYjmFpfKAJCYOURtP1t_l8AvNn/view?usp=sharing), which show the estimates match well. This was only trained from a few hours, so we can further reduce the error with more epochs. You can reduce the learning rate further if the loss saturates. I've saved all the weights from the three epochs [here](https://drive.google.com/drive/folders/1S5sr9d0V1XCgJnAORULjEKue82DJkuq4?usp=sharing). With 7px validation error after just 3 epochs, this is quite lower than what you reported from your attempt. 

---
4. To check how the embeddings look, I used the TDN weights to create a codebook for the sugar_box object. I ran the following to get some rough results, this can ofcourse be made better with further epochs of TDN training: 
- `python midastouch/tactile_tree/build_codebook.py expt.obj_model=004_sugar_box expt.codebook_size=50000`
- `python /home/suddhu/Projects/MidasTouch/midastouch/eval/viz_codebook.py`

![rsz_screenshot_2023-02-01_at_114114_pm](https://user-images.githubusercontent.com/20095927/216233459-98af3685-c90f-49b9-80a7-fc7c10b7798c.png)

---

**TL;DR**: I followed the same data generation and training pipeline as in the open-source repo, using a larger learning rate for initial convergence. Let me know if you have any further questions, and thanks for your interest!

**Comment by gabrielecaddeo** [comment]
Hi @suddhu ,

Thank you very much for your time! I will try it and then I will let you know.

**Comment by gabrielecaddeo** [comment]
Hi @suddhu,
I tested the last commit, it worked! We can consider this issue closed, thank you!
